### PR TITLE
create ssl cert per vhost, not one monster

### PR DESCRIPTION
**Pull request by evgeni**

### CODE REVIEW

File: puppet/modules/web/manifests/init.pp [hunk 1]
```diff
@@ -15,37 +15,6 @@
     letsencrypt => $https,
```
Comment (ekohl) [inline puppet/modules/web/manifests/init.pp:15]: I'd prefer to drop this parameter and move the inclusion into the vhost:
https://github.com/theforeman/foreman-infra/blob/cc1300e15cbb07e3b94ded6aaa53f0fbe77132fe/puppet/modules/web/manifests/base.pp#L8-L10

File: puppet/modules/web/manifests/vhost.pp [hunk 2]
```diff
@@ -52,6 +52,12 @@
   }
 
   if $web::https {
```
Comment (ekohl) [inline puppet/modules/web/manifests/vhost.pp:54]: As noted above, let's include the prerequisite here
```suggestion
  if $web::https {
    include web::letsencrypt
```

**Review by ekohl — CHANGES_REQUESTED** [review]
tl;dr: :+1: on the concept, some implementation notes inline.

I recently read there's some benefit in having all names on a certificate so browsers can reuse connections, but given we're mostly behind a CDN I think the benefit of that is really limited and the operational simplicity of one cert per vhost is way more important.

In the installer we use [datacat](https://github.com/richardc/puppet-datacat) to gather all values so you could do it that way per host, but it's complex. Also, the module is unmaintained and [collections](https://github.com/voxpupuli/puppet-collections) is a maintained alternative.

File: puppet/modules/web/manifests/init.pp (hunk 1 above)
Comment (evgeni) [reply puppet/modules/web/manifests/init.pp:15]: That will break using `web::base` with HTTPS directly, like it's done by `redmine` and `profiles::jenkins::controller`, as those do not use `web::vhost`.

**Comment by ehelms** [comment]
This solution works for me as an approach. Thanks for taking a look.

File: puppet/modules/web/manifests/init.pp (hunk 1 above)
Comment (ekohl) [reply puppet/modules/web/manifests/init.pp:15]: Can we modify those to also include the base? I think b31f6de8158d629c1ba7751a84a07b2db5650789 should work and makes it all more consistent.

File: puppet/modules/web/manifests/init.pp [hunk 3]
```diff
@@ -15,37 +15,6 @@
...
   }
 
-  if $https {
-    $letsencypt_domain = 'theforeman.org'
```
Comment (ekohl) [inline puppet/modules/web/manifests/init.pp:19]: Looks like there was a typo here, but at least it was used consistently.

**Comment by evgeni** [comment]
finally updated!

**Review by ehelms — APPROVED** [review]
LGTM, but @ekohl should give the final stamp of approval

File: puppet/modules/web/manifests/jenkins.pp
Comment (ekohl) [inline puppet/modules/web/manifests/jenkins.pp:1]: Looks like we could migrate this to using `web::vhost` as well after this change.

https://webmasters.stackexchange.com/questions/97005/setting-x-forwarded-proto-under-apache-2-4 suggests you can also use `RequestHeader set X-Forwarded-Proto expr=%{REQUEST_SCHEME}` to make it dynamic, meaning it's going to be the same directive for HTTP and HTTPS.

I can submit a follow up PR to do so.

Comment (evgeni) [reply puppet/modules/web/manifests/jenkins.pp:1]: please :)

Comment (ekohl) [reply puppet/modules/web/manifests/jenkins.pp:1]: https://github.com/theforeman/foreman-infra/pull/2144
